### PR TITLE
Fix createMany action hooks

### LINKED ISSUE #16198: items.create action hooks triggered by `createMany` cannot query inserted items
### Describe the Bug

Action hook handlers on `some_collection.items.create` cannot query newly created records when they are inserted through `createMany`, because the hooks are triggered by `createOne` that runs in a nested, [unfinished transaction](https://github.com/directus/directus/blob/d6fce8d92a48a5983293ca0c04f9de84201ec405/api/src/services/items.ts#L263). Trying to query the inserted record using `readOne` will yield an "You don't have permission to access this." error because the transaction has not been committed yet. The only exception being the last record in the `createMany` batch.

I am aware that querying the created record is not needed in `items.create` hooks because they get the data in `metadata.payload`, but sometimes it is feasible to use the same handler for both `items.update` and `items.create` hooks. Also, the handler might need more data not available in the payload, such as db defaults for unset fields or related data from foreign key tables.

Note that this is not an issue when records are inserted using `createOne` because there is no nested transaction.

### To Reproduce

* add an action hook `action('my_collection.items.create', handler)`
* insert multiple records into my_collection using `createMany`
* Let `handler` query the record with the inserted primary key given in `metadata.key`
  * using `readOne`: `const row = await itemService.readOne(metadata.key)` --> throws "You don't have permission to access this."
  * using `readMany`: `const rows = await itemService.readMany([metadata.key]);` --> `rows.length === 0`

### Errors Shown

"You don't have permission to access this." when using `readOne`

### What version of Directus are you using?

9.18.1

### What version of Node.js are you using?

16

### What database are you using?

Postgres 13

### What browser are you using?

Chrome

### How are you deploying Directus?

docker

### PR DESCRIPTION
**Pull request by appy-one**
## Description

Fixes `items.create` action hooks not being able to query inserted items when created through `createMany`. Instead of emitting the events from `createOne`, it queues them using `bypassEmitAction`, then emits them once the `createMany`'s transaction has been finished/committed.

In addition (see 2nd commit): nested O2M/A2O/M20 events queued by payloadService were emitted in `createOne` and `updateMany`, regardless of a local `opts.bypassEmitAction` being provided, causing the same issue for create and update action hooks on related data being created/updated. Those hooks could potentially query old/non-existent data, because their parent's transaction hasn't been committed yet. With this commit, all nestedActionEvents will be queued and bubble up the callstack until no `bypassEmitAction` has been set - then emitted.

I haven't added test code yet, will follow shortly.

Fixes #16198

## Type of Change

- [x] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

File: api/src/services/items.ts [hunk 1]
```diff
@@ -260,23 +264,36 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
 	 * Create multiple new items at once. Inserts all provided records sequentially wrapped in a transaction.
 	 */
 	async createMany(data: Partial<Item>[], opts?: MutationOptions): Promise<PrimaryKey[]> {
-		const primaryKeys = await this.knex.transaction(async (trx) => {
+		const { primaryKeys, emitActions } = await this.knex.transaction(async (trx) => {
```
Comment (licitdev) [inline api/src/services/items.ts:267]: Can we rename `emitActions` to `nestedActionEvents` for consistency with other similar usages? Another similar renaming in line 289 from actionEvent into nestedActionEvent.

Comment (appy-one) [reply api/src/services/items.ts]: ✔️

**Comment by appy-one** [comment]
I discovered the same hooks issue is present when importing data: both json and csv imports call `upsertOne` for each row within a transaction without passing a `bypassEmitAction` argument. Any `create` and `update` hooks triggered by this will not be able to query the new/updated data. I'm pushing another commit to fix this.

**Comment by appy-one** [comment]
@licitdev @rijkvanzanten I tried to produce additional tests but run into "Mock handler not found" errors as soon as I set `emitEvents: true` on any insert. Can someone take a look at this, would be great if the PR gets merged soon so we get reliable hooks!

**Comment by licitdev** [comment]
@appy-one This PR was pending https://github.com/directus/directus/pull/15576 which had just been merged. I'll be re-reviewing this PR soon. Feel free to push up your additional tests, even if they're failing with `emitEvents: true`. Thank you for your help! 👍

**Comment by appy-one** [comment]
@licitdev when will you be able to review?

**Comment by licitdev** [comment]
@appy-one Ah... I thought you were adding some tests, so the review was held back. I'll look into this soon. Thanks for the heads up.

**Comment by licitdev** [comment]
@appy-one Please allow access for maintainers on your PR so that we can push changes to your branch.

image

You've mentioned about running into issue with mocking when writing tests, mind checking out the tests in #16433? 
I think it might be a good reference and we should look into adding tests for this PR.

**Comment by appy-one** [comment]
@licitdev I have no such option
Screenshot 2022-12-15 at 13 31 12

**Comment by licitdev** [comment]
@appy-one Ah... I wonder if it's because this PR is a fork of an organisation. 
Would you like to attempt the tests or would you prefer for me to take over the branch in a new PR? 🤔

**Comment by appy-one** [comment]
@licitdev sorry for the late reply - feel free to take over. Co-authored comment would be appreciated

**Comment by licitdev** [comment]
Superseded by #17066. 
@appy-one Your commits are pushed there and you should be a co-author. Thank you.

**Comment by rickyitexpert** [comment]
This Bug is also present in upsertMany Method.

**Comment by licitdev** [comment]
@rickyitexpert Kindly open a new bug report. Thank you.

**Pull request closed without merging**